### PR TITLE
Make everything `position: relative`

### PR DESCRIPTION
**Pull request by Andy-set-studio**
This is something @davidkpiano talks about and at first I was _not on board_ and even made fun of him about it, but the more I think about it, the more I like it. 

## Possible benefits

- Predictable parent relationships for `absolute` children
- Stacking becomes easier (yes I know there's a plethora of ways that you can affect stacking order)
- The potential of a lot less CSS written. The chances of removing `relative` for `static` etc are likely to be _way_ lower that you’d add `relative`

I’ve been thinking about it again when [he and Shaw were on ShopTalkShow](https://shoptalkshow.com/episodes/381/), which is worth a listen.

It’s worth remembering that a lot of stuff in this reset goes against the browser defaults, so this one, albeit _way_ against browser defaults fits in nicely IMO. 

I’m opening this up for discussion because I might be missing a big issue here which I really don't want to create.

### CODE REVIEW

**Comment by Andy-set-studio** [comment]
Thread on Twitter: https://twitter.com/hankchizljaw/status/1182396654001053696

**Comment by emilio-martinez** [comment]
I've deployed this type of change in a fairly large project, e.g., ~2 million page site. It worked well, so I'd say this is a fair inclusion.

**Comment by yatil** [comment]
With the caveat that I don’t use CSS resets (yes, I’m THAT kind of guy). When I use absolute positioning, it is a red flag that says “be aware what you are doing here because that can cause problems with zoom, responsiveness, overlapping and so on”.

Therefore absolute positioning is something I only use when I need to position something that can’t be done using grid/flex positioning. It makes those cases very rare and often means that I position something from a nested element in relationship to an element that is at least one layer up.

It feels counter-intuitive to me to then remove position: relative until I get to where I want to anchor the element to.

TL;DR: When I set position:absolute deliberately, I also set position:relative deliberately.

**Comment by hidde** [comment]
Adding this will make it more fiddly to intentionally choose what context you're absolutely positioning a given element against. If everything is relatively positioned, positioning context will always be the parent (unless one resets the parent's position). I think it is fairly common to want to position against a (great)grandparent, and would therefore advise against this.

**Comment by davidkpiano** [comment]
> Adding this will make it more fiddly to intentionally choose what context you're absolutely positioning a given element against

Perhaps this is a good constraint?

> I think it is fairly common to want to position against a (great)grandparent, and would therefore advise against this

No, this is very uncommon. 

**To prove this:**

1. Go to any webpage with non-trivial styles (such as this one, on GitHub)
2. Add the `* { position: relative }` style via browser dev tools
3. See if any styling breaks.

Chances are, very few things (or nothing at all) will break.

**Comment by Andy-set-studio** [comment]
Reply for @hidde:

I think in that context, a rule like this could be feasible:

```css
.component-with-absolute-grandchild,
.component-with-absolute-grandchild * {
  position: static;
}

.component-with-absolute-grandchild .absolute-thing {
  position: absolute;
}
```

That's presuming that this context is a rarity in your codebase though. If this is frequently having to be dealt with then I'd personally remove this rule line in the reset, just like I’d remove the nuclear reduced-motion rules if I had a responsible contextual system in place.

**Comment by hidde** [comment]
@davidkpiano 

> Perhaps this is a good constraint?

If it was part of the CSS spec, maybe. But it is a constraint that exists just on websites that use this reset, and not on others. I worry this will make it confusing for people who have learned CSS, as they now need to unlearn some of their expectations of how it works. Basically it changes how the standard works for only some websites.

For example: developer works on website A, website B, website C. Then they're hired by an agency to tweak some styles on website D, which uses this reset. Now positioning works differently than expected. I realise doing `^ { box-sizing: border-box; }`, which I do like, potentially has the same problem.

> Chances are, very few things (or nothing at all) will break. 

This does surprise me! But it does not take away my worries… the “few things” might include something that is super hard to debug / that your whole business relies on / that causes hard-to-automatically-detect accessibility issues for some users etc

@hankchizljaw Right, yea, it is easy to re-reset. But if it wasn't there in the first place, it wouldn't need resetting… I guess the question really is, like you and @davidkpiano say, whether it is a rarity or not. To me it doesn't feel rare, but understand for many it would certainly be rare.

**Comment by Andy-set-studio** [comment]
The beauty of this reset is that it's so tiny, taking elements away that are causing problems is really easy. A few people don’t like the `ul[class], ol[class]` stuff and my response to them each time is basically “don't use them, then”.

**Comment by Andy-set-studio** [comment]
Closing this. I've used this PR version for a couple of prototypes and small projects and if I’m honest, it didn’t really improve things for _me_.

I think the concept is solid though: I just think for this project, maybe it's not the right thing to add.
